File: cloud-layer/cloud-feed-service/src/services/kpiService.ts
import { PrismaClient, Prisma } from '@prisma/client'
import { logger } from '../utils/logger'
import { Decimal } from '@prisma/client/runtime/library'

const prisma = new PrismaClient()

// Weight data from telemetry service (for MVP, we call via HTTP)
interface TelemetryReading {
  id: string
  tenantId: string
  farmId: string | null
  barnId: string | null
  deviceId: string
  batchId: string | null
  metric: string
  value: number | string
  unit: string | null
  occurredAt: string
}

interface TelemetryAggregate {
  id: string
  tenantId: string
  farmId: string | null
  barnId: string | null
  deviceId: string
  metric: string
  bucketStart: string
  bucketSize: string
  avgValue: number | string
  minValue: number | string
  maxValue: number | string
  count: number
}

interface DailyWeightData {
  date: string
  avgWeight: number | null
  weightGain: number | null
}

/**
 * Fetch weight data from telemetry service
 * TODO: For production, consider caching or direct DB access for performance
 */
async function fetchWeightData(params: {
  tenantId: string
  barnId: string
  batchId?: string | null
  startDate: Date
  endDate: Date
  headers?: Record<string, string>
}): Promise<DailyWeightData[]> {
  const telemetryBaseUrl =
    process.env.TELEMETRY_BASE_URL || 'http://cloud-telemetry-service:3000'

  try {
    // Try to get daily aggregates first (more efficient)
    const aggUrl = new URL(`${telemetryBaseUrl}/api/v1/telemetry/aggregates`)
    aggUrl.searchParams.set('tenantId', params.tenantId)
    aggUrl.searchParams.set('barnId', params.barnId)
    if (params.batchId) aggUrl.searchParams.set('batchId', params.batchId)
    aggUrl.searchParams.set('metric', 'weight')
    aggUrl.searchParams.set('from', params.startDate.toISOString())
    aggUrl.searchParams.set('to', params.endDate.toISOString())
    aggUrl.searchParams.set('bucket', '1d')

    const aggResponse = await fetch(aggUrl.toString(), {
      method: 'GET',
      headers: {
        'content-type': 'application/json',
        ...(params.headers || {}),
      },
    })

    if (aggResponse.ok) {
      const aggregates = (await aggResponse.json()) as TelemetryAggregate[]
      // Convert aggregates to daily weight data
      const dailyWeights: Map<string, { avgWeight: number | null; weightGain: number | null }> =
        new Map()

      // Group by date and compute average weight per day
      aggregates.forEach((agg) => {
        const date = new Date(agg.bucketStart).toISOString().split('T')[0]
        const avgValue = typeof agg.avgValue === 'string' ? parseFloat(agg.avgValue) : agg.avgValue

        if (!dailyWeights.has(date)) {
          dailyWeights.set(date, { avgWeight: avgValue, weightGain: null })
        } else {
          const existing = dailyWeights.get(date)!
          existing.avgWeight = avgValue
        }
      })

      // Compute weight gain (today - yesterday)
      const sortedDates = Array.from(dailyWeights.keys()).sort()
      sortedDates.forEach((date, idx) => {
        if (idx > 0) {
          const prevDate = sortedDates[idx - 1]
          const prevWeight = dailyWeights.get(prevDate)?.avgWeight
          const currWeight = dailyWeights.get(date)?.avgWeight

          if (prevWeight !== null && prevWeight !== undefined && currWeight !== null && currWeight !== undefined) {
            const gain = currWeight - prevWeight
            dailyWeights.get(date)!.weightGain = gain > 0 ? gain : null
          }
        }
      })

      // Return as array sorted by date (with date key for mapping)
      return sortedDates.map((date) => ({
        date,
        avgWeight: dailyWeights.get(date)!.avgWeight,
        weightGain: dailyWeights.get(date)!.weightGain,
      })) as DailyWeightData[]
    }

    // Fallback to raw readings if aggregates not available
    const readingsUrl = new URL(`${telemetryBaseUrl}/api/v1/telemetry/readings`)
    readingsUrl.searchParams.set('tenantId', params.tenantId)
    readingsUrl.searchParams.set('barnId', params.barnId)
    if (params.batchId) readingsUrl.searchParams.set('batchId', params.batchId)
    readingsUrl.searchParams.set('metric', 'weight')
    readingsUrl.searchParams.set('from', params.startDate.toISOString())
    readingsUrl.searchParams.set('to', params.endDate.toISOString())
    readingsUrl.searchParams.set('limit', '10000')

    const readingsResponse = await fetch(readingsUrl.toString(), {
      method: 'GET',
      headers: {
        'content-type': 'application/json',
        ...(params.headers || {}),
      },
    })

    if (!readingsResponse.ok) {
      logger.warn('Failed to fetch weight data from telemetry service', {
        status: readingsResponse.status,
      })
      return []
    }

    const readings = (await readingsResponse.json()) as TelemetryReading[]

    // Group readings by date and compute daily averages
    const dailyMap = new Map<string, number[]>()
    readings.forEach((reading) => {
      const date = new Date(reading.occurredAt).toISOString().split('T')[0]
      const value = typeof reading.value === 'string' ? parseFloat(reading.value) : reading.value

      if (!dailyMap.has(date)) {
        dailyMap.set(date, [])
      }
      dailyMap.get(date)!.push(value)
    })

    // Compute daily averages and weight gains
    const dailyWeights: DailyWeightData[] = []
    const sortedDates = Array.from(dailyMap.keys()).sort()

    sortedDates.forEach((date, idx) => {
      const values = dailyMap.get(date) || []
      const avgWeight = values.length > 0 ? values.reduce((a, b) => a + b, 0) / values.length : null

      let weightGain: number | null = null
      if (idx > 0) {
        const prevDate = sortedDates[idx - 1]
        const prevValues = dailyMap.get(prevDate) || []
        const prevAvg =
          prevValues.length > 0 ? prevValues.reduce((a, b) => a + b, 0) / prevValues.length : null

        if (prevAvg !== null && avgWeight !== null && avgWeight > prevAvg) {
          weightGain = avgWeight - prevAvg
        }
      }

      dailyWeights.push({ date: sortedDates[idx], avgWeight, weightGain })
    })

    return dailyWeights
  } catch (error) {
    logger.error('Error fetching weight data from telemetry service', error)
    return []
  }
}

/**
 * Compute KPI series for a date range
 * Option A (MVP): Compute on-demand per request
 */
export async function computeKpiSeries(params: {
  tenantId: string
  barnId: string
  batchId?: string | null
  startDate: Date
  endDate: Date
  headers?: Record<string, string>
}) {
  const { tenantId, barnId, batchId, startDate, endDate } = params

  try {
    // 1. Fetch feed intake records grouped by date
    const intakeRecords = await prisma.feedIntakeRecord.findMany({
      where: {
        tenantId,
        barnId,
        ...(batchId ? { batchId } : {}),
        occurredAt: {
          gte: startDate,
          lte: endDate,
        },
      },
      select: {
        quantityKg: true,
        occurredAt: true,
      },
    })

    // Group intake by date
    const dailyIntake = new Map<string, number>()
    intakeRecords.forEach((record) => {
      const date = record.occurredAt.toISOString().split('T')[0]
      const quantity = parseFloat(record.quantityKg.toString())
      dailyIntake.set(date, (dailyIntake.get(date) || 0) + quantity)
    })

    // 2. Fetch weight data from telemetry service
    const weightData = await fetchWeightData({
      tenantId,
      barnId,
      batchId,
      startDate,
      endDate,
      headers: params.headers,
    })

    // 3. Get animal count data (for ADG calculation)
    // TODO: Get from barn-records-service when available
    // For MVP, assume constant count or fetch from batch table if available
    let animalCount: number | null = null
    if (batchId) {
      // Try to get animal count from tenant-registry batch table
      // For MVP, we'll use a default or fetch from a future barn-records integration
      // For now, we'll compute ADG without animal count (per-animal gain if count available)
    }

    // 4. Get mortality data for adjustments
    // TODO: Get from barn-records-service when available
    const mortalityAdjustments = new Map<string, number>()

    // 5. Compute daily KPIs
    const kpiItems: Array<{
      recordDate: string
      fcr: number | null
      adgG: number | null
      sgrPct: number | null
    }> = []

    // Build a map of all dates in range
    const allDates = new Set<string>()
    Array.from(dailyIntake.keys()).forEach((date: string) => allDates.add(date))
    weightData.forEach((wd) => {
      allDates.add(wd.date)
    })
    // Also add all dates in the range (for completeness)
    const currentDate = new Date(startDate)
    while (currentDate <= endDate) {
      allDates.add(currentDate.toISOString().split('T')[0])
      currentDate.setDate(currentDate.getDate() + 1)
    }

    const sortedDates = Array.from(allDates).sort()

    // Build weight data map by date for quick lookup
    const weightDataMap = new Map<string, DailyWeightData>()
    weightData.forEach((wd) => {
      weightDataMap.set(wd.date, wd)
    })

    let prevWeight: number | null = null

    sortedDates.forEach((date) => {
      const feedKg = dailyIntake.get(date) || 0
      const weightInfo = weightDataMap.get(date) || { date, avgWeight: null, weightGain: null }

      const avgWeight = weightInfo.avgWeight
      let weightGainKg = weightInfo.weightGain
      const mortalityCount = mortalityAdjustments.get(date) || 0

      // Apply mortality adjustment to weight gain
      // Rule: If animal count decreases, weight gain should be adjusted
      // For MVP: simple proportional adjustment
      if (mortalityCount > 0 && weightGainKg !== null && animalCount !== null) {
        // Adjust: subtract weight contribution from dead animals
        // Simplified: assume dead animals had average weight
        const avgWeightPerAnimal = avgWeight && animalCount > 0 ? avgWeight / animalCount : null
        if (avgWeightPerAnimal !== null) {
          weightGainKg = weightGainKg - avgWeightPerAnimal * mortalityCount
        }
      }

      // Compute FCR
      // Corner case: weight_gain <= 0 => FCR = null
      let fcr: number | null = null
      if (feedKg > 0 && weightGainKg !== null && weightGainKg > 0) {
        fcr = feedKg / weightGainKg
      } else if (feedKg > 0 && (weightGainKg === null || weightGainKg <= 0)) {
        // Corner case: weight_gain <= 0 or missing
        fcr = null // Flagged by weightMissingFlag or weightGainNonPositive
      }
      // Corner case: missing intake => omit FCR (intakeMissingFlag = true)

      // Compute ADG (Average Daily Gain in grams per day)
      // ADG = weight_gain_kg / animal_count (or per animal if count available)
      let adgG: number | null = null
      if (weightGainKg !== null && weightGainKg > 0) {
        if (animalCount !== null && animalCount > 0) {
          // Per-animal ADG
          adgG = (weightGainKg / animalCount) * 1000 // Convert to grams
        } else {
          // Total ADG (no per-animal breakdown available)
          adgG = weightGainKg * 1000
        }
      }

      // Compute SGR (Specific Growth Rate in %)
      // SGR = ((ln(Wt) - ln(W0)) / days) * 100
      let sgrPct: number | null = null
      if (prevWeight !== null && avgWeight !== null && prevWeight > 0 && avgWeight > 0) {
        const days = 1 // Daily rollup
        sgrPct = ((Math.log(avgWeight) - Math.log(prevWeight)) / days) * 100
      }

      kpiItems.push({
        recordDate: date,
        fcr,
        adgG,
        sgrPct,
      })

      if (avgWeight !== null) {
        prevWeight = avgWeight
      }
    })

    // 6. Optionally upsert into kpi_daily for caching (MVP: skip for now, compute on-demand)

    return {
      items: kpiItems,
    }
  } catch (error) {
    logger.error('Error computing KPI series', error)
    throw error
  }
}

File: cloud-layer/cloud-analytics-service/app/kpi_service.py
from __future__ import annotations

import logging
from datetime import date, datetime
from typing import Any, Optional

import httpx

from app.config import Settings
from app.db import AnalyticsDb

logger = logging.getLogger(__name__)


class FeedingKpiService:
    def __init__(self, db: AnalyticsDb, settings: Settings):
        self.db = db
        self.settings = settings
        self.feed_service_url = settings.feed_service_url or "http://cloud-feed-service:5130"
        self.barn_records_service_url = settings.barn_records_service_url or "http://cloud-barn-records-service:3000"
        self.weighvision_readmodel_url = settings.weighvision_readmodel_url or "http://cloud-weighvision-readmodel:3000"

    async def get_weight_aggregates(
        self,
        tenant_id: str,
        farm_id: Optional[str],
        barn_id: str,
        batch_id: Optional[str],
        start: date,
        end: date,
    ) -> dict[date, dict[str, Any]]:
        """Fetch weight aggregates from weighvision-readmodel or barn-records fallback."""
        result: dict[date, dict[str, Any]] = {}

        # Try weighvision-readmodel first
        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                params = {
                    "tenant_id": tenant_id,
                    "start": start.isoformat(),
                    "end": end.isoformat(),
                }
                if farm_id:
                    params["farm_id"] = farm_id
                if barn_id:
                    params["barn_id"] = barn_id
                if batch_id:
                    params["batch_id"] = batch_id

                resp = await client.get(
                    f"{self.weighvision_readmodel_url}/api/v1/weighvision/weight-aggregates",
                    params=params,
                )
                if resp.status_code == 200:
                    data = resp.json()
                    if isinstance(data, dict) and "items" in data:
                        for item in data["items"]:
                            date_str = item.get("date") or item.get("record_date")
                            if not date_str:
                                continue
                            try:
                                record_date = datetime.fromisoformat(date_str.replace("Z", "+00:00")).date()
                            except (ValueError, AttributeError):
                                # Try parsing as date string directly
                                from datetime import datetime as dt
                                try:
                                    record_date = dt.strptime(date_str, "%Y-%m-%d").date()
                                except ValueError:
                                    continue
                            result[record_date] = {
                                "avg_weight_kg": item.get("avg_weight_kg"),
                                "p10": item.get("p10") or item.get("p10_weight_kg"),
                                "p50": item.get("p50") or item.get("p50_weight_kg"),
                                "p90": item.get("p90") or item.get("p90_weight_kg"),
                                "sample_count": item.get("sample_count", 0),
                                "quality_pass_rate": item.get("quality_pass_rate", 1.0),
                                "source": "weighvision",
                            }
                    return result
        except Exception as e:
            logger.warning(f"Failed to fetch from weighvision-readmodel: {e}")

        # Fallback to barn-records daily counts
        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                params = {
                    "tenant_id": tenant_id,
                    "start": start.isoformat(),
                    "end": end.isoformat(),
                }
                if farm_id:
                    params["farm_id"] = farm_id
                if barn_id:
                    params["barn_id"] = barn_id
                if batch_id:
                    params["batch_id"] = batch_id

                resp = await client.get(
                    f"{self.barn_records_service_url}/api/v1/barn-records/daily-counts",
                    params=params,
                )
                if resp.status_code == 200:
                    data = resp.json()
                    if isinstance(data, dict) and "items" in data:
                        for item in data["items"]:
                            date_str = item.get("recordDate") or item.get("record_date")
                            if not date_str:
                                continue
                            try:
                                record_date = datetime.fromisoformat(date_str.replace("Z", "+00:00")).date()
                            except (ValueError, AttributeError):
                                # Try parsing as date string directly
                                from datetime import datetime as dt
                                try:
                                    record_date = dt.strptime(date_str, "%Y-%m-%d").date()
                                except ValueError:
                                    continue
                            if record_date not in result:
                                result[record_date] = {
                                    "avg_weight_kg": item.get("averageWeightKg") or item.get("avg_weight_kg"),
                                    "source": "barn_records",
                                }
        except Exception as e:
            logger.warning(f"Failed to fetch from barn-records: {e}")

        return result

    async def get_feed_intake(
        self,
        tenant_id: str,
        farm_id: Optional[str],
        barn_id: str,
        batch_id: Optional[str],
        start: date,
        end: date,
    ) -> dict[date, float]:
        """Fetch feed intake records and aggregate by date."""
        result: dict[date, float] = {}

        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                params = {
                    "tenantId": tenant_id,
                    "start": start.isoformat(),
                    "end": end.isoformat(),
                }
                if farm_id:
                    params["farmId"] = farm_id
                if barn_id:
                    params["barnId"] = barn_id
                if batch_id:
                    params["batchId"] = batch_id

                resp = await client.get(
                    f"{self.feed_service_url}/api/v1/feed/intake-records",
                    params=params,
                )
                if resp.status_code == 200:
                    data = resp.json()
                    if isinstance(data, dict) and "items" in data:
                        for item in data["items"]:
                            occurred_at_str = item.get("occurredAt") or item.get("occurred_at")
                            if not occurred_at_str:
                                continue
                            try:
                                occurred_at = datetime.fromisoformat(occurred_at_str.replace("Z", "+00:00"))
                                record_date = occurred_at.date()
                            except (ValueError, AttributeError):
                                continue
                            quantity_kg = float(item.get("quantityKg") or item.get("quantity_kg") or 0)
                            result[record_date] = result.get(record_date, 0) + quantity_kg
        except Exception as e:
            logger.warning(f"Failed to fetch feed intake: {e}")

        return result

    async def get_daily_counts(
        self,
        tenant_id: str,
        farm_id: Optional[str],
        barn_id: str,
        batch_id: Optional[str],
        start: date,
        end: date,
    ) -> dict[date, dict[str, Any]]:
        """Fetch daily counts (animal count, mortality, etc.) from barn-records."""
        result: dict[date, dict[str, Any]] = {}

        try:
            async with httpx.AsyncClient(timeout=5.0) as client:
                params = {
                    "tenant_id": tenant_id,
                    "start": start.isoformat(),
                    "end": end.isoformat(),
                }
                if farm_id:
                    params["farm_id"] = farm_id
                if barn_id:
                    params["barn_id"] = barn_id
                if batch_id:
                    params["batch_id"] = batch_id

                resp = await client.get(
                    f"{self.barn_records_service_url}/api/v1/barn-records/daily-counts",
                    params=params,
                )
                if resp.status_code == 200:
                    data = resp.json()
                    if isinstance(data, dict) and "items" in data:
                        for item in data["items"]:
                            date_str = item.get("recordDate") or item.get("record_date")
                            if not date_str:
                                continue
                            try:
                                record_date = datetime.fromisoformat(date_str.replace("Z", "+00:00")).date()
                            except (ValueError, AttributeError):
                                # Try parsing as date string directly
                                from datetime import datetime as dt
                                try:
                                    record_date = dt.strptime(date_str, "%Y-%m-%d").date()
                                except ValueError:
                                    continue
                            result[record_date] = {
                                "animal_count": item.get("animalCount") or item.get("animal_count"),
                                "mortality_count": item.get("mortalityCount") or item.get("mortality_count") or 0,
                                "cull_count": item.get("cullCount") or item.get("cull_count") or 0,
                            }
        except Exception as e:
            logger.warning(f"Failed to fetch daily counts: {e}")

        return result

    async def compute_kpi_for_date(
        self,
        tenant_id: str,
        farm_id: Optional[str],
        barn_id: str,
        batch_id: Optional[str],
        target_date: date,
        weight_data: dict[date, dict[str, Any]],
        feed_data: dict[date, float],
        daily_counts: dict[date, dict[str, Any]],
    ) -> Optional[dict[str, Any]]:
        """Compute KPI for a single date."""
        weight_info = weight_data.get(target_date, {})
        feed_kg = feed_data.get(target_date, 0)
        counts = daily_counts.get(target_date, {})

        animal_count = counts.get("animal_count")
        avg_weight_kg = weight_info.get("avg_weight_kg")

        if not animal_count or animal_count <= 0:
            return None

        if not avg_weight_kg or avg_weight_kg <= 0:
            avg_weight_kg = None

        # Compute biomass
        biomass_kg = None
        if avg_weight_kg and animal_count:
            biomass_kg = avg_weight_kg * animal_count

        # Compute weight gain (need previous day)
        weight_gain_kg = None
        from datetime import timedelta
        prev_date = target_date - timedelta(days=1)
        if prev_date:
            prev_weight = weight_data.get(prev_date, {}).get("avg_weight_kg")
            if prev_weight and avg_weight_kg:
                prev_biomass = prev_weight * animal_count
                if biomass_kg:
                    weight_gain_kg = biomass_kg - prev_biomass

        # Compute FCR
        fcr = None
        if weight_gain_kg and weight_gain_kg > 0 and feed_kg > 0:
            fcr = feed_kg / weight_gain_kg

        # Compute ADG (Average Daily Gain) in grams
        adg_kg = None
        if weight_gain_kg and animal_count:
            adg_kg = (weight_gain_kg / animal_count) * 1000  # Convert to grams

        # Compute SGR (Specific Growth Rate) in percent
        sgr_pct = None
        if avg_weight_kg and prev_weight and prev_weight > 0:
            import math

            days = 1
            sgr_pct = ((math.log(avg_weight_kg) - math.log(prev_weight)) / days) * 100

        # Flags
        intake_missing_flag = feed_kg == 0
        weight_missing_flag = avg_weight_kg is None
        quality_flag = not (intake_missing_flag or weight_missing_flag)

        return {
            "tenant_id": tenant_id,
            "farm_id": farm_id,
            "barn_id": barn_id,
            "batch_id": batch_id,
            "date": target_date,
            "animal_count": animal_count,
            "avg_weight_kg": avg_weight_kg,
            "biomass_kg": biomass_kg,
            "weight_gain_kg": weight_gain_kg,
            "total_feed_kg": feed_kg,
            "fcr": fcr,
            "adg_kg": adg_kg,
            "sgr_pct": sgr_pct,
            "intake_missing_flag": intake_missing_flag,
            "weight_missing_flag": weight_missing_flag,
            "quality_flag": quality_flag,
        }

    async def upsert_kpi(self, kpi_data: dict[str, Any]) -> None:
        """Upsert KPI record into feeding_kpi_daily table."""
        assert self.db._pool is not None
        async with self.db._pool.acquire() as conn:
            await conn.execute(
                """
                INSERT INTO feeding_kpi_daily (
                  tenant_id, farm_id, barn_id, batch_id, date,
                  animal_count, avg_weight_kg, biomass_kg, weight_gain_kg,
                  total_feed_kg, fcr, adg_kg, sgr_pct,
                  intake_missing_flag, weight_missing_flag, quality_flag,
                  updated_at
                )
                VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, now())
                ON CONFLICT (tenant_id, barn_id, batch_id, date) DO UPDATE
                SET
                  animal_count = EXCLUDED.animal_count,
                  avg_weight_kg = EXCLUDED.avg_weight_kg,
                  biomass_kg = EXCLUDED.biomass_kg,
                  weight_gain_kg = EXCLUDED.weight_gain_kg,
                  total_feed_kg = EXCLUDED.total_feed_kg,
                  fcr = EXCLUDED.fcr,
                  adg_kg = EXCLUDED.adg_kg,
                  sgr_pct = EXCLUDED.sgr_pct,
                  intake_missing_flag = EXCLUDED.intake_missing_flag,
                  weight_missing_flag = EXCLUDED.weight_missing_flag,
                  quality_flag = EXCLUDED.quality_flag,
                  updated_at = now()
                """,
                kpi_data["tenant_id"],
                kpi_data["farm_id"],
                kpi_data["barn_id"],
                kpi_data["batch_id"],
                kpi_data["date"],
                kpi_data["animal_count"],
                kpi_data["avg_weight_kg"],
                kpi_data["biomass_kg"],
                kpi_data["weight_gain_kg"],
                kpi_data["total_feed_kg"],
                kpi_data["fcr"],
                kpi_data["adg_kg"],
                kpi_data["sgr_pct"],
                kpi_data["intake_missing_flag"],
                kpi_data["weight_missing_flag"],
                kpi_data["quality_flag"],
            )

    async def get_kpi_series(
        self,
        tenant_id: str,
        farm_id: Optional[str],
        barn_id: str,
        batch_id: Optional[str],
        start: date,
        end: date,
    ) -> list[dict[str, Any]]:
        """Get KPI series from materialized table."""
        assert self.db._pool is not None
        where = ["tenant_id = $1", "barn_id = $2", "date >= $3", "date <= $4"]
        args: list[Any] = [tenant_id, barn_id, start, end]
        idx = 5

        if farm_id:
            where.append(f"farm_id = ${idx}")
            args.append(farm_id)
            idx += 1
        if batch_id:
            where.append(f"batch_id = ${idx}")
            args.append(batch_id)
            idx += 1

        sql = f"""
          SELECT
            date, animal_count, avg_weight_kg, biomass_kg, weight_gain_kg,
            total_feed_kg, fcr, adg_kg, sgr_pct,
            intake_missing_flag, weight_missing_flag, quality_flag
          FROM feeding_kpi_daily
          WHERE {" AND ".join(where)}
          ORDER BY date ASC
        """

        async with self.db._pool.acquire() as conn:
            rows = await conn.fetch(sql, *args)

        return [
            {
                "recordDate": str(row["date"]),
                "animalCount": row["animal_count"],
                "avgWeightKg": row["avg_weight_kg"],
                "biomassKg": row["biomass_kg"],
                "weightGainKg": row["weight_gain_kg"],
                "totalFeedKg": row["total_feed_kg"],
                "fcr": row["fcr"],
                "adgG": row["adg_kg"],
                "sgrPct": row["sgr_pct"],
                "intakeMissingFlag": row["intake_missing_flag"],
                "weightMissingFlag": row["weight_missing_flag"],
                "qualityFlag": row["quality_flag"],
            }
            for row in rows
        ]

